rfcs/0159-private-task-logs.md

```diff
@@ -0,0 +1,66 @@
+
+This RFC proposes making this behavior optional, with private logs as an alternative.
+Such logs would require credentials with specific scopes to read.
+
+## Motivation
+
+Even in open-source development, secrecy can be important.
+For example, CI for fixes to security issues can reveal those security issues if logs are public.
+
+Private-log capability is also a common request from outside parties interested in deploying Taskcluster.
+
+# Details
+
+## Current Log Implementation
+
+The current [convention](https://docs.taskcluster.net/docs/manual/design/conventions/task-logs) is to name the main task log `public/logs/live.log`, and that fetching this artifact may -- if the worker supports it -- result in receiving log lines as they are produced.
+
+The Taskcluster UI, and other consumers of task logs such as Treeherder, incorporate this convention and look for logs at that artifact name.
+
+Workers that cannot do "live logging" simply create an S3 artifact with this name.
+Workers that *can* do live logging use the following process.
+At the beginning of the task, they create a Reference artifact named `public/logs/live.log` containing a URL that will connect to the worker (usually via [Websocktunnel](https://github.com/taskcluster/taskcluster/tree/master/tools/websocktunnel#readme)) to get log lines as they are produced.
+When the task is complete, they create a new S3 artifact containing the whole log, named `public/logs/live_backing.log`.
+They then modify the URL in the `public/logs/live.log` artifact (using a [special case in the `createTask` API](https://docs.taskcluster.net/docs/reference/platform/queue/api#createArtifact)) to point to the URL for the `getArtifact` method for the `public/logs/live_backing.log` artifact (`<rootUrl>/api/queue/v1/task/<taskId>/runs/<runId>/artifacts/public/logs/live_backing.log`).
+
+Reference artifacts correspond to HTTP 303 redirects, and HTTP clients do not re-authenticate when following a redirect.
+So the fetch of `live_backing.log` is performed without Taskcluster credentials, even if the initial call to `getArtifact` used credentials.
+If `live_backing.log` is always public, this is not a problem, but if it is private then this may result in additional complexity for clients wishing to access task logs.
+
+## New Log Implementation
+
+The updated convention is, briefly, to make the artifact name containing logs configurable on a per-task basis, so that users may choose a `public/` or non-public name; and to introduce a new artifact type that will avoid the issues around authentication and redirects.
+
+The convention for task-related conventions is to put information in `task.extra`, using sub-objects to group related configuration.
+In the new design, the log artiact name is found at `task.extra.logs.taskLogArtifact`, with a default value of `public/logs/live.log` for compatibility.
```

Comment:
Taking point 2 first, I think there's a strong sense among users of CI systems that there's a single "main" log that should be displayed by default.  That doesn't preclude having other logs, but those can be named whatever the user would like and handled outside of the convention for the main logs (what this proposal calls "task logs").  For example, if there are both stderr.log and stdout.log, then the user would click on the one they want to see.

There's a related issue of how UIs can figure out which artifacts (aside from the main log) are logs and which are files to be downloaded, but I think that's out of scope for this RFC.  The current approach (based on name, I think?) seems good enough.

To the first point, I understand the purity in the idea that the worker *only* looks at `task.payload`.  But the practical downside to this is that the user must then coordinate two values in the same task JSON, with a failure to do so correctly resulting in no logs appearing.  From a user perspective that doesn't make a lot of sense.  It also means that the user has to figure out how to configure that differently for each worker.

Workers are pretty practical beasts -- they already look at various other bits of the task such as deadline, expires, and scopes, and they log lots of maybe-useful information automatically.

Perhaps a compromise would be allowing configuration of the task log artifact name in the payload, but defaulting that to `task.extra.logs.taskLogArtifact`, which in turn defaults to `public/logs/live.log`?

It would be nice to have the default be `task.log`, but I think in the interests of compatibility we'll need to stick with `live.log`.  The impact of logs suddenly "disappearing" for some group of users -- or the cost of carefully coordinating a migration to a new default -- would probably be worse than the impact of wondering why it's named `live.log`.

Reply:
Defaulting to public will make migrating easier but will be a potentially surprising default later in time. I am thinking it would be better to ensure that current tasks we want to be public all update their task.extra before we land this change and default to private logs.

The nice thing about this is that the fallout if we make some tasks logs private by accident isn't that bad, we just use our scopes to go get the artifact for the interested party and update their config so that it Just Works in the future.

Reply:
What's the advantage?  Why would that be a surprising default? Are people surprised by that behavior in GitLab or Travis-CI?

Making logs private by default changes this from a relatively consequence-free change to a "contact everyone who uses Taskcluster and get them to spend their precious time migrating" project -- a substantially more difficult task and similar to a few dozen half-finished migrations we have lying around.

Reply:
I like the idea of there being a single log artefact, as this was slightly surprising to me when I initially started learning about TaskCluster. Is there any reason that TC currently doesn't allow replacing artefacts after creation?

Possibly not a concern for now, but could we keep the log location in `payload` if we had some means at task creation to pass arbitrary metadata and for this task pass something which to begin with just updates the field in `extra`? (As an aside, arbitrary metadata would answer the download question since you could include 'Content-Disposition', which both S3 and GCS support setting on objects.)

Reply:
> I like the idea of there being a single log artefact, as this was slightly surprising to me when I initially started learning about TaskCluster. Is there any reason that TC currently doesn't allow replacing artefacts after creation?

This is an issue of data integrity -- we want to ensure that an artifact containing, say, a release binary doesn't change after it's created.  It's also part of the general immutable design of Taskcluster.  The exception for changing Reference artifacts hasn't proven problematic, so I think we could expand on that with LiveLog.

I'm not sure I understand the second paragraph, @ricky26.  A task, once created, is immutable (similar reasons), so there's no way to "update" anything.  So `task.extra` and `task.payload` must both be fully formed at the call to `queue.createTask`.

That's a good point about content-disposition, and something we should think about in the Object Service (#155).  I think we would want something visible in `listArtifacts` so that the UI could display logs and files differently, meaning that it's metadata in the queue rather than the target storage system.  But for direct links to artifacts, content-distribution would definitely help the browser know what to do.

Reply:
Aha, I didn't realise that replacing the reference assets was a special case! I follow the reasoning now, thanks for explaining.

Reply:
> What's the advantage? Why would that be a surprising default? Are people surprised by that behavior in GitLab or Travis-CI?

I would be surprised if I was a relatively unfamiliar user and had been using taskcluster in a project that had private logs that when I went to go set up a new project I had to set this one subfield of a subfield in a task definition to keep my logs from being public.